File: APP&WEB/public_html/V2/_archive/old-scripts/scripts/generate_invoice.py
#!/usr/bin/env python3
"""
ZION eShop - Invoice Generator CLI
===================================
Generuje PDF faktury z příkazové řádky pro integraci s PHP.

Usage:
    python3 scripts/generate_invoice.py \
        --invoice-number "2025/001" \
        --order-id "ORD123456" \
        --customer-name "Jan Novák" \
        --customer-email "jan@example.com" \
        --customer-address "Ulice 123, Praha" \
        --items '[{"name":"Product 1","quantity":2,"unit_price":500}]' \
        --output-path "/path/to/invoice.pdf"

Author: ZION Team
Created: 2025-12-09
"""

import sys
import os
import json
import argparse
from pathlib import Path
from datetime import datetime, timedelta

# Add correct project src to sys.path (works under public_html and server root)
script_path = Path(__file__).resolve()
public_html_root = script_path.parents[2]  # .../public_html
server_root = public_html_root.parent      # .../newearth.cz

# Prefer server_root/src if exists, else fallback to public_html/src
src_candidates = [
    server_root / "src",
    public_html_root / "src",
]
for cand in src_candidates:
    if cand.exists():
        sys.path.insert(0, str(cand))
        break

# Optional override via env var
env_src = os.environ.get("ZION_SRC_PATH")
if env_src:
    sys.path.insert(0, env_src)

from wallet.rasta_invoice_generator import (
    RastaInvoiceGenerator,
    InvoiceData,
    InvoiceItem,
    CompanyInfo
)


def parse_items(items_json: str) -> list:
    """Parse items from JSON string.
    
    Args:
        items_json: JSON array of items
        
    Returns:
        List of InvoiceItem objects
    """
    try:
        items_data = json.loads(items_json)
        items = []
        for item_data in items_data:
            items.append(InvoiceItem(
                name=item_data['name'],
                quantity=int(item_data['quantity']),
                unit_price=float(item_data['unit_price']),
                vat_rate=float(item_data.get('vat_rate', 0.21))
            ))
        return items
    except Exception as e:
        print(f"❌ Error parsing items: {e}", file=sys.stderr)
        sys.exit(1)


def main():
    """Main CLI entry point."""
    parser = argparse.ArgumentParser(description='Generate ZION Rasta-themed invoice PDF')
    
    # Invoice details
    parser.add_argument('--invoice-number', required=True, help='Invoice number (e.g., 2025/001)')
    parser.add_argument('--order-id', required=True, help='Order ID')
    parser.add_argument('--issue-date', help='Issue date (YYYY-MM-DD, default: today)')
    parser.add_argument('--due-date', help='Due date (YYYY-MM-DD, default: +14 days)')
    
    # Customer info
    parser.add_argument('--customer-name', required=True, help='Customer name')
    parser.add_argument('--customer-email', required=True, help='Customer email')
    parser.add_argument('--customer-address', required=True, help='Customer address')
    parser.add_argument('--customer-ico', help='Customer IČO')
    parser.add_argument('--customer-dic', help='Customer DIČ')
    
    # Items
    parser.add_argument('--items', required=True, help='Items JSON array')
    
    # Payment
    parser.add_argument('--payment-method', default='Bankovní převod', help='Payment method')
    parser.add_argument('--variable-symbol', help='Variable symbol (default: invoice number)')
    
    # Output
    parser.add_argument('--output-path', required=True, help='Output PDF path')
    parser.add_argument('--logo-path', help='Path to company logo (optional)')
    
    # Optional
    parser.add_argument('--notes', help='Additional notes')
    
    # Currency (presale support)
    parser.add_argument("--currency", default="CZK", help="Currency code (CZK, EUR)")
    parser.add_argument("--exchange-rate", type=float, default=25.0, help="EUR to CZK exchange rate")
    parser.add_argument("--dual-currency", type=int, default=0, help="Show both currencies (0 or 1)")
    
    args = parser.parse_args()
    
    # Parse dates
    issue_date = args.issue_date or datetime.now().strftime('%Y-%m-%d')
    due_date = args.due_date or (datetime.now() + timedelta(days=14)).strftime('%Y-%m-%d')
    
    # Parse items
    items = parse_items(args.items)
    
    # Create invoice data
    invoice = InvoiceData(
        invoice_number=args.invoice_number,
        order_id=args.order_id,
        issue_date=issue_date,
        due_date=due_date,
        customer_name=args.customer_name,
        customer_address=args.customer_address,
        customer_ico=args.customer_ico,
        customer_dic=args.customer_dic,
        items=items,
        payment_method=args.payment_method,
        variable_symbol=args.variable_symbol,
        notes=args.notes
    )
    
    # Generate invoice
    try:
        generator = RastaInvoiceGenerator(logo_path=args.logo_path)
        output_path = generator.generate(invoice, args.output_path)
        
        # Success output (JSON for PHP parsing)
        result = {
            'success': True,
            'output_path': output_path,
            'invoice_number': invoice.invoice_number,
            'total': invoice.total,
            'total_formatted': f"{invoice.total:,.2f} Kč"
        }
        print(json.dumps(result, ensure_ascii=False))
        sys.exit(0)
        
    except Exception as e:
        # Error output (JSON)
        error = {
            'success': False,
            'error': str(e)
        }
        print(json.dumps(error, ensure_ascii=False), file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: APP&WEB/public_html/V2/generate-invoice.php
<?php
/**
 * ZION eShop - Invoice Generator PHP Wrapper
 * ============================================
 * Generuje PDF fakturu z objednávky pomocí Python backendu.
 *
 * Usage:
 *     $result = generateInvoice($order);
 *     if ($result['success']) {
 *         $invoicePath = $result['output_path'];
 *     }
 *
 * Author: ZION Team
 * Created: 2025-12-09
 */

/**
 * Generate PDF invoice from order data
 *
 * @param array $order Order data from create-order.php
 * @return array Result with success, output_path, and error fields
 */
require_once __DIR__ . '/php-python-bridge.php';

function generateInvoice(array $order): array
{
    // Basic validation
    if (empty($order['orderId']) || empty($order['customer'])) {
        return ['success' => false, 'error' => 'Missing required order fields'];
    }

    // Paths - na serveru je struktura: public_html/V2/scripts/, public_html/V2/invoices/
    $pythonScript = __DIR__ . '/../scripts/generate_invoice.py';
    $invoicesDir = __DIR__ . '/../invoices';
    
    // Create invoices directory if not exists
    if (!is_dir($invoicesDir)) {
        mkdir($invoicesDir, 0755, true);
    }
    
    // Generate invoice number from order ID and date
    $invoiceNumber = date('Y') . '/' . str_replace('ORD-', '', $order['orderId']);
    $outputPath = $invoicesDir . '/invoice_' . $order['orderId'] . '.pdf';
    
    // Prepare customer data
    $customer = $order['customer'];
    $customerName = $customer['name'] ?? '';
    $customerEmail = $customer['email'] ?? '';
    
    // Build customer address
    $addressParts = [];
    $custAddr = $customer['address'] ?? null;
    if (is_array($custAddr)) {
        if (!empty($custAddr['street'])) $addressParts[] = $custAddr['street'];
        if (!empty($custAddr['city'])) $addressParts[] = $custAddr['city'];
        if (!empty($custAddr['zip'])) $addressParts[] = $custAddr['zip'];
        if (!empty($custAddr['country'])) $addressParts[] = $custAddr['country'];
    } else {
        if (!empty($customer['street'])) $addressParts[] = $customer['street'];
        if (!empty($customer['city'])) $addressParts[] = $customer['city'];
        if (!empty($customer['zip'])) $addressParts[] = $customer['zip'];
        if (!empty($customer['country'])) $addressParts[] = $customer['country'];
    }
    
    // If shipping to Zásilkovna pickup point, use that address
    if (!empty($order['shipping']['pickupPoint'])) {
        $pp = $order['shipping']['pickupPoint'];
        $addressParts = [
            $pp['name'] ?? '',
            ($pp['street'] ?? '') . ', ' . ($pp['city'] ?? ''),
            $pp['zip'] ?? ''
        ];
    }
    
    $customerAddress = implode("\n", array_filter($addressParts));
    
    // Prepare items JSON
    $items = [];
    if (!empty($order['items']) && is_array($order['items'])) {
        foreach ($order['items'] as $item) {
            $items[] = [
                'name' => $item['name'] ?? $item['title'] ?? 'Unknown Item',
                'quantity' => intval($item['quantity'] ?? 1),
                'unit_price' => floatval($item['price'] ?? 0),
                'vat_rate' => 0.21
            ];
        }
    } elseif (($order['type'] ?? '') === 'presale' && !empty($order['package']['priceEur'])) {
        // Presale fallback: vytvořit jednu položku dle balíčku, bez DPH (0%) v EUR
        $pkgName = 'ZION Presale - ' . ($order['package']['name'] ?? 'Balíček');
        $items[] = [
            'name' => $pkgName,
            'quantity' => 1,
            'unit_price' => floatval($order['package']['priceEur']),
            'vat_rate' => 0.0
        ];
    } else {
        return ['success' => false, 'error' => 'No items provided'];
    }
    
    // Add shipping as item if applicable
    if (!empty($order['shipping']['price']) && $order['shipping']['price'] > 0) {
        $shippingName = 'Doprava: ' . ($order['shipping']['method'] ?? 'Zásilkovna');
        $items[] = [
            'name' => $shippingName,
            'quantity' => 1,
            'unit_price' => floatval($order['shipping']['price']),
            'vat_rate' => 0.21
        ];
    }
    
    $itemsJson = json_encode($items, JSON_UNESCAPED_UNICODE);
    
    // Dates
    $issueDate = date('Y-m-d');
    $dueDate = date('Y-m-d', strtotime('+14 days'));

    // Currency: presale v EUR, eShop v CZK (Kč)
    $isPresale = (($order['type'] ?? '') === 'presale');
    $currency = $isPresale ? 'EUR' : 'CZK';
    // Exchange rate for dual display (optional)
    $exchangeRate = 25.0; // default fallback
    if (defined('EXCHANGE_RATE_EUR_CZK')) {
        $exchangeRate = floatval(EXCHANGE_RATE_EUR_CZK);
    }
    
    // Payment method (avoid match for compatibility)
    $payment = $order['payment'] ?? [];
    $pmRaw = is_array($payment) ? ($payment['method'] ?? 'bank_transfer') : (string)$payment;
    $pm = strtolower($pmRaw);
    if (in_array($pm, ['bank_transfer', 'bank', 'prevod'])) {
        $paymentMethod = 'Bankovní převod';
    } elseif (in_array($pm, ['card', 'karta'])) {
        $paymentMethod = 'Platební karta';
    } elseif (in_array($pm, ['cash', 'hotovost'])) {
        $paymentMethod = 'Hotově';
    } elseif (in_array($pm, ['crypto', 'bitcoin'])) {
        $paymentMethod = 'Kryptoměny';
    } else {
        $paymentMethod = 'Bankovní převod';
    }
    
    // Variable symbol = extract numeric timestamp from orderId
    // PRESALE-1767340840-81b95d → 1767340840
    // ORD-1767340840 → 1767340840
    if (preg_match('/(\d{10})/', $order['orderId'], $vsMatch)) {
        $variableSymbol = $vsMatch[1];
    } else {
        $variableSymbol = preg_replace('/[^0-9]/', '', $order['orderId']);
    }
    
    // Build CLI args for bridge (no shell)
    $cliArgs = [
        '--invoice-number' => $invoiceNumber,
        '--order-id' => $order['orderId'],
        '--issue-date' => $issueDate,
        '--due-date' => $dueDate,
        '--customer-name' => $customerName,
        '--customer-email' => $customerEmail,
        '--customer-address' => $customerAddress,
        '--items' => $itemsJson,
        '--payment-method' => $paymentMethod,
        '--variable-symbol' => $variableSymbol,
        '--output-path' => $outputPath,
        '--currency' => $currency,
        '--exchange-rate' => strval($exchangeRate),
        '--dual-currency' => $isPresale ? '1' : '0'
    ];

    // Explicitní logo path (bypass autodetekce v Pythonu)
    // Primárně použij cestu relativní k tomuto souboru: ../img/logo144.png
    $logoPath = realpath(__DIR__ . '/../img/logo144.png');
    if ($logoPath === false || !file_exists($logoPath)) {
        // Možné absolutní cesty na hostingu (fallbacky)
        $fallbacks = [
            '/home/html/newearth.cz/public_html/V2/img/logo144.png',
            '/home/html/newearth.cz/public_html/images/logo144.png',
        ];
        foreach ($fallbacks as $fb) {
            if (file_exists($fb)) {
                $logoPath = $fb;
                break;
            }
        }
    }
    if ($logoPath && file_exists($logoPath)) {
        $cliArgs['--logo-path'] = $logoPath;
    }
    
    // Add IČO/DIČ if present
    if (!empty($customer['ico'])) {
        $cliArgs['--customer-ico'] = $customer['ico'];
    }
    if (!empty($customer['dic'])) {
        $cliArgs['--customer-dic'] = $customer['dic'];
    }

    // Execute Python script via bridge (bypass shell)
    $pythonPath = '/usr/bin/python3';
    $result = runPythonScriptWithCLI($pythonPath, $pythonScript, $cliArgs);
    
    // Debug log
    $debugLog = __DIR__ . '/../logs/invoice-generation-debug.log';
    $debugData = date('Y-m-d H:i:s') . " - Order: {$order['orderId']}\n";
    $debugData .= "Exit Code: " . ($result['exit_code'] ?? 'NULL') . "\n";
    $debugData .= "STDOUT: " . ($result['stdout'] ?? 'NULL') . "\n";
    $debugData .= "STDERR: " . ($result['stderr'] ?? 'NULL') . "\n";
    if (!empty($cliArgs['--logo-path'])) {
        $debugData .= "LOGO: " . $cliArgs['--logo-path'] . "\n";
    } else {
        $debugData .= "LOGO: (auto)\n";
    }
    $debugData .= str_repeat('-', 80) . "\n";
    file_put_contents($debugLog, $debugData, FILE_APPEND);
    
    if (($result['exit_code'] ?? 1) === 0) {
        // Try to parse JSON output from stdout
        $outputStr = trim($result['stdout'] ?? '');
        $parsed = json_decode($outputStr, true);
        
        if ($parsed && isset($parsed['success']) && $parsed['success']) {
            return [
                'success' => true,
                'output_path' => $parsed['output_path'],
                'invoice_number' => $parsed['invoice_number'],
                'total' => $parsed['total'] ?? 0,
                'total_formatted' => $parsed['total_formatted'] ?? '',
                'path' => $parsed['output_path'] // Alias pro kompatibilitu
            ];
        }
    }
    
    // Error occurred
    $errorOutput = trim(($result['stdout'] ?? '') . "\n" . ($result['stderr'] ?? ''));
    error_log("Invoice generation failed for order {$order['orderId']}: {$errorOutput}");
    
    return [
        'success' => false,
        'error' => 'Failed to generate invoice',
        'output' => $errorOutput,
        'exit_code' => $result['exit_code'] ?? -1,
        'args' => $cliArgs
    ];
}

/**
 * Get invoice URL for email attachment or download
 *
 * @param string $invoicePath Full filesystem path to invoice PDF
 * @return string Public URL to invoice
 */
function getInvoiceUrl(string $invoicePath): string
{
    // Convert filesystem path to URL
    // /path/to/public_html/V2/invoices/invoice_XXX.pdf -> /V2/invoices/invoice_XXX.pdf
    $relativePath = str_replace(
        dirname(__DIR__, 1),
        '',
        $invoicePath
    );
    
    return 'https://newearth.cz' . $relativePath;
}
